src/protocol/clock/period.rs

```diff
@@ -0,0 +1,108 @@
+use std::num::{IntErrorKind, TryFromIntError};
+use std::time::Duration;
+
+pub type PeriodLength = Duration;
+pub type PeriodCount = u64;
+pub type PeriodTotalTime = Duration;
+
+pub trait Periods: Sized + Default {
+    fn new(length: &PeriodLength, count: &PeriodCount) -> Self;
+
+    fn add_count(&self, add: PeriodCount) -> Result<Self, IntErrorKind>;
+    fn sub_count(&self, sub: PeriodCount) -> Result<Self, IntErrorKind>;
+}
+
+pub trait PeriodsTotals: Periods {
+    fn total_time(&self) -> Result<Option<PeriodTotalTime>, TryFromIntError>;
```

Comment:
Nested Results or Options such as `Result<Option<T>, E>` and `Option<Result<T, E>>` are not really nice to work with. We can avoid the parent Result type by changing `Period.count` to `u32`. Else maybe create a custom error enum and return `Result<PeriodTotalTime, PeriodError>` instead. :)

Reply:
Hmm. I thought that it is better to just pass through the error to the caller to deal with it there. (if they want to).